### PR TITLE
En 4972 unit tests fork detector

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**

### CODE REVIEW

File: process/sync/baseForkDetector_test.go [hunk 1]
```diff
@@ -291,21 +175,89 @@ func TestBasicForkDetector_CheckForkOnlyOneShardHeaderOnANonceShouldReturnFalse(
...
+		false,
+	)
+	forkInfo := bfd.CheckFork()
+	assert.False(t, forkInfo.IsDetected)
+	assert.Equal(t, uint64(math.MaxUint64), forkInfo.Nonce)
+	assert.Nil(t, forkInfo.Hash)
+}
+
+func TestBasicForkDetector_CheckForkOnlyReceivedHeadersShouldReturnFalse(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/baseForkDetector_test.go]: add empty line

File: process/sync/baseForkDetector_test.go [hunk 2]
```diff
@@ -291,21 +175,89 @@ func TestBasicForkDetector_CheckForkOnlyOneShardHeaderOnANonceShouldReturnFalse(
...
+		false,
+	)
+	forkInfo := bfd.CheckFork()
+	assert.False(t, forkInfo.IsDetected)
+	assert.Equal(t, uint64(math.MaxUint64), forkInfo.Nonce)
+	assert.Nil(t, forkInfo.Hash)
+}
+
+func TestBasicForkDetector_CheckForkOnlyOneShardHeaderOnANonceReceivedAndProcessedShouldReturnFalse(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/baseForkDetector_test.go]: add empty line

File: process/sync/metaForkDetector_test.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	assert.NotNil(t, sfd)
+
+	assert.Equal(t, uint64(0), sfd.LastCheckpointNonce())
+	assert.Equal(t, uint64(0), sfd.LastCheckpointRound())
+	assert.Equal(t, uint64(0), sfd.FinalCheckpointNonce())
+	assert.Equal(t, uint64(0), sfd.FinalCheckpointRound())
+}
+
+func TestMetaForkDetector_AddHeaderNilHeaderShouldErr(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/metaForkDetector_test.go]: add empty line

File: process/sync/metaForkDetector_test.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	rounderMock := &mock.RounderMock{RoundIndex: 100}
+	bfd, _ := sync.NewMetaForkDetector(rounderMock, &mock.BlackListHandlerStub{})
+	err := bfd.AddHeader(nil, make([]byte, 0), process.BHProcessed, nil, nil, false)
+	assert.Equal(t, sync.ErrNilHeader, err)
+}
+
+func TestMetaForkDetector_AddHeaderNilHashShouldErr(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/metaForkDetector_test.go]: add empty line

File: process/sync/metaForkDetector_test.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	rounderMock := &mock.RounderMock{RoundIndex: 100}
+	bfd, _ := sync.NewMetaForkDetector(rounderMock, &mock.BlackListHandlerStub{})
+	err := bfd.AddHeader(&block.Header{}, nil, process.BHProcessed, nil, nil, false)
+	assert.Equal(t, sync.ErrNilHash, err)
+}
+
+func TestMetaForkDetector_AddHeaderUnsignedBlockShouldErr(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/metaForkDetector_test.go]: add empty line

File: process/sync/metaForkDetector_test.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+		make([]byte, 0),
+		process.BHProcessed,
+		nil,
+		nil,
+		false)
+	assert.Equal(t, sync.ErrBlockIsNotSigned, err)
+}
+
+func TestMetaForkDetector_AddHeaderNotPresentShouldWork(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/metaForkDetector_test.go]: add empty line

File: process/sync/metaForkDetector_test.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+	err := bfd.AddHeader(hdr, hash, process.BHProcessed, nil, nil, false)
+	assert.Nil(t, err)
+
+	hInfos := bfd.GetHeaders(1)
+	assert.Equal(t, 1, len(hInfos))
+	assert.Equal(t, hash, hInfos[0].Hash())
+}
+
+func TestMetaForkDetector_AddHeaderPresentShouldAppend(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/metaForkDetector_test.go]: add empty line

File: process/sync/metaForkDetector_test.go [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+	hash2 := []byte("hash2")
+	rounderMock := &mock.RounderMock{RoundIndex: 1}
+	bfd, _ := sync.NewMetaForkDetector(rounderMock, &mock.BlackListHandlerStub{
+		HasCalled: func(key string) bool {
+			return false
+		},
+	})
+
+	_ = bfd.AddHeader(hdr1, hash1, process.BHProcessed, nil, nil, false)
+	err := bfd.AddHeader(hdr2, hash2, process.BHProcessed, nil, nil, false)
```
Comment (bogdan-rosianu) [inline process/sync/metaForkDetector_test.go]: why do you check for error only second time ?

File: process/sync/metaForkDetector_test.go [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+	assert.Nil(t, err)
+
+	hInfos := bfd.GetHeaders(1)
+	assert.Equal(t, 2, len(hInfos))
+	assert.Equal(t, hash1, hInfos[0].Hash())
+	assert.Equal(t, hash2, hInfos[1].Hash())
+}
+
+func TestMetaForkDetector_AddHeaderWithProcessedBlockShouldSetCheckpoint(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/metaForkDetector_test.go]: add empty line

File: process/sync/metaForkDetector_test.go [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+		HasCalled: func(key string) bool {
+			return false
+		},
+	})
+	_ = bfd.AddHeader(hdr1, hash1, process.BHProcessed, nil, nil, false)
+	assert.Equal(t, hdr1.Nonce, bfd.LastCheckpointNonce())
+}
+
+func TestMetaForkDetector_AddHeaderPresentShouldNotRewriteState(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/metaForkDetector_test.go]: add empty line

File: process/sync/metaForkDetector_test.go [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+	hInfos := bfd.GetHeaders(1)
+	assert.Equal(t, 2, len(hInfos))
+	assert.Equal(t, hash, hInfos[0].Hash())
+	assert.Equal(t, process.BHReceived, hInfos[0].GetBlockHeaderState())
+	assert.Equal(t, process.BHProcessed, hInfos[1].GetBlockHeaderState())
+}
+
+func TestMetaForkDetector_AddHeaderHigherNonceThanRoundShouldErr(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/metaForkDetector_test.go]: add empty line

File: process/sync/shardForkDetector_test.go [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+	assert.NotNil(t, sfd)
+
+	assert.Equal(t, uint64(0), sfd.LastCheckpointNonce())
+	assert.Equal(t, uint64(0), sfd.LastCheckpointRound())
+	assert.Equal(t, uint64(0), sfd.FinalCheckpointNonce())
+	assert.Equal(t, uint64(0), sfd.FinalCheckpointRound())
+}
+
+func TestShardForkDetector_AddHeaderNilHeaderShouldErr(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/shardForkDetector_test.go]: add empty line

File: process/sync/shardForkDetector_test.go [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+	rounderMock := &mock.RounderMock{RoundIndex: 100}
+	bfd, _ := sync.NewShardForkDetector(rounderMock, &mock.BlackListHandlerStub{})
+	err := bfd.AddHeader(nil, make([]byte, 0), process.BHProcessed, nil, nil, false)
+	assert.Equal(t, sync.ErrNilHeader, err)
+}
+
+func TestShardForkDetector_AddHeaderNilHashShouldErr(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/shardForkDetector_test.go]: add empty line

File: process/sync/shardForkDetector_test.go [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+	rounderMock := &mock.RounderMock{RoundIndex: 100}
+	bfd, _ := sync.NewShardForkDetector(rounderMock, &mock.BlackListHandlerStub{})
+	err := bfd.AddHeader(&block.Header{}, nil, process.BHProcessed, nil, nil, false)
+	assert.Equal(t, sync.ErrNilHash, err)
+}
+
+func TestShardForkDetector_AddHeaderUnsignedBlockShouldErr(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/shardForkDetector_test.go]: add empty line

File: process/sync/shardForkDetector_test.go [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+		make([]byte, 0),
+		process.BHProcessed,
+		nil,
+		nil,
+		false)
+	assert.Equal(t, sync.ErrBlockIsNotSigned, err)
+}
+
+func TestShardForkDetector_AddHeaderNotPresentShouldWork(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/shardForkDetector_test.go]: add empty line

File: process/sync/shardForkDetector_test.go [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+	hash2 := []byte("hash2")
+	rounderMock := &mock.RounderMock{RoundIndex: 1}
+	bfd, _ := sync.NewShardForkDetector(rounderMock, &mock.BlackListHandlerStub{
+		HasCalled: func(key string) bool {
+			return false
+		},
+	})
+
+	_ = bfd.AddHeader(hdr1, hash1, process.BHProcessed, nil, nil, false)
+	err := bfd.AddHeader(hdr2, hash2, process.BHProcessed, nil, nil, false)
```
Comment (bogdan-rosianu) [inline process/sync/shardForkDetector_test.go]: also here. why do you check only the second time for error?

File: process/sync/shardForkDetector_test.go [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+	assert.Nil(t, err)
+
+	hInfos := bfd.GetHeaders(1)
+	assert.Equal(t, 2, len(hInfos))
+	assert.Equal(t, hash1, hInfos[0].Hash())
+	assert.Equal(t, hash2, hInfos[1].Hash())
+}
+
+func TestShardForkDetector_AddHeaderWithProcessedBlockShouldSetCheckpoint(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/shardForkDetector_test.go]: add empty line

File: process/sync/shardForkDetector_test.go [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+		HasCalled: func(key string) bool {
+			return false
+		},
+	})
+	_ = bfd.AddHeader(hdr1, hash1, process.BHProcessed, nil, nil, false)
+	assert.Equal(t, hdr1.Nonce, bfd.LastCheckpointNonce())
+}
+
+func TestShardForkDetector_AddHeaderPresentShouldNotRewriteState(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/shardForkDetector_test.go]: add empty line

File: process/sync/shardForkDetector_test.go [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+
+	hInfos := bfd.GetHeaders(1)
+	assert.Equal(t, 2, len(hInfos))
+	assert.Equal(t, hash, hInfos[0].Hash())
+	assert.Equal(t, process.BHReceived, hInfos[0].GetBlockHeaderState())
+	assert.Equal(t, process.BHProcessed, hInfos[1].GetBlockHeaderState())
+}
+
+func TestShardForkDetector_AddHeaderHigherNonceThanRoundShouldErr(t *testing.T) {
+	t.Parallel()
```
Comment (bogdan-rosianu) [inline process/sync/shardForkDetector_test.go]: add empty line

File: process/sync/metaForkDetector_test.go (hunk 8 above)
Comment (BeniaminDrasovean) [reply process/sync/metaForkDetector_test.go]: I add a header with the same parameters twice. I know that the first time the func will not return an err, but I want to check that if I add the same header a second time, the err will be nil.

File: process/sync/shardForkDetector_test.go (hunk 16 above)
Comment (BeniaminDrasovean) [reply process/sync/shardForkDetector_test.go]: Same answer as before.

**Pull request merged**